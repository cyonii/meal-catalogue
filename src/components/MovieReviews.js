import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ListGroup from 'react-bootstrap/ListGroup';
import { fetchReviews } from '../api/movies';

const MovieReviews = (props) => {
  const { movieID } = props;
  const [reviews, setReviews] = useState([]);

  useEffect(() => {
    fetchReviews(movieID).then((data) => setReviews(data));
  }, []);

  return (
    <>
      <h5 className="sticky-top bg-dark py-2 text-decoration-underline border-secondary">Reviews</h5>
      <ListGroup variamt="flush">
        {
          reviews.length > 0
            ? reviews.map((review) => (
              <ListGroup.Item className="px-0 bg-transparent text-white" key={review.id}>
                <h6 className="text-info mb-0 lh-1">{`${review.author}`}</h6>
                <small className="text-muted">{`Rating: ${review.author_details.rating}`}</small>

                {review.content.split('\n').map((line) => (<p className="mb-2 opacity-75" key={Math.random()}>{line}</p>))}
              </ListGroup.Item>
            ))
            : (
              <ListGroup.Item className="px-0 bg-transparent text-center mt-3 mt-md-5">
                <h6 className="display-6 my-auto fw-bold mb-0 lh-1">No reviews yet.</h6>
              </ListGroup.Item>
            )
        }
      </ListGroup>
    </>
  );
};

MovieReviews.propTypes = {
  movieID: PropTypes.number.isRequired,
};

export default MovieReviews;
