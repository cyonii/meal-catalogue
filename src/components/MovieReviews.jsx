import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ListGroup from 'react-bootstrap/ListGroup';
import { moviedb } from '../api/movies';
import ReviewItem from './ReviewItem';
import Loading from './widgets/Loading';

const MovieReviews = (props) => {
  const { movieID } = props;
  const [reviews, setReviews] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(async () => {
    setLoading(true);
    await moviedb.movieReviews(movieID)
      .then((data) => setReviews(data.results))
      .catch((err) => err);
    setLoading(false);
  }, []);

  return (
    <>
      <h5 className="sticky-top bg-dark py-2 text-decoration-underline border-secondary">Reviews</h5>
      { loading
        ? <Loading size="1rem" />
        : (
          <ListGroup as="ul" variant="flush" data-testid="review-list">
            { reviews.length < 1
              ? <h2 className="my-auto opacity-25 fw-bold mb-0 lh-1">No reviews yet.</h2>
              : reviews.map((review) => <ReviewItem review={review} key={review.id} />)}
          </ListGroup>
        )}
    </>
  );
};

MovieReviews.propTypes = {
  movieID: PropTypes.number.isRequired,
};

export default MovieReviews;
