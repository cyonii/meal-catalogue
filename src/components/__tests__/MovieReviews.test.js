import {
  screen, render, waitFor, cleanup,
} from '@testing-library/react';
import MovieReviews from '../MovieReviews';
import reviewData from './data/reviews.json';

const mockReviews = reviewData;

// mock moviedb.movieReviews api calls
jest.mock('../../api/movies', () => ({
  moviedb: {
    movieReviews: () => Promise.resolve(mockReviews),
  },
}));

describe('MovieReviews', () => {
  afterEach(cleanup);

  it('should render Loading component while fetching reviews', async () => {
    render(<MovieReviews movieID={56778} />);

    await waitFor(() => {
      expect(screen.getByTestId('loading')).toBeInTheDocument();
      expect(screen.getByTestId('loading')).toMatchSnapshot();
    });
  });

  it('should render review list items', async () => {
    render(<MovieReviews movieID={56778} />);

    await waitFor(() => {
      expect(screen.getByTestId('review-list')).toMatchSnapshot();
      expect(screen.getByTestId('review-list')).toBeInTheDocument();
      expect(screen.getAllByTestId('review-item').length).toBe(3);
    });
  });
});
