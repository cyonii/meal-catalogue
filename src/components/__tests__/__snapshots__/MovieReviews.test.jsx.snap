// Vitest Snapshot v1

exports[`MovieReviews > should render Loading component while fetching reviews 1`] = `
<div
  class="d-flex flex-column justify-content-center align-items-center mt-5"
  data-testid="loading"
>
  <i
    class="bi bi-lightning-charge text-secondary spin"
    style="font-size: 1rem;"
  />
</div>
`;

exports[`MovieReviews > should render review list items 1`] = `
<ul
  class="list-group list-group-flush"
  data-testid="review-list"
>
  <li
    class="px-0 bg-transparent text-white border-0 border-bottom border-secondary list-group-item"
    data-testid="review-item"
  >
    <h6
      class="text-info mb-0 lh-1"
    >
      Cat Ellington
    </h6>
    <div
      class="d-flex justify-content-between align-items-center mb-3"
    >
      <small
        class="text-muted"
      >
        Rating: null
      </small>
    </div>
    <p
      class="mb-2 opacity-75 review-text"
    >
      (As I'm writing this review, Darth Vader's theme music begins to build in my mind...)
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      Well, it actually has a title, what the Darth Vader theme. And that title is "The Imperial March", composed by the great John Williams, whom, as many of you may 
      <span
        class="text-muted fw-bold"
      >
        ...
      </span>
    </p>
    <div
      class="d-flex mt-1 opacity-100"
    >
      <button
        class="lh-1 py-0 pb-1 rounded-0 text-decoration-none btn btn-info"
        type="button"
      >
        <small
          class="fw-bold"
          style="font-size: 13px;"
        >
          Show more
        </small>
      </button>
    </div>
  </li>
  <li
    class="px-0 bg-transparent text-white border-0 border-bottom border-secondary list-group-item"
    data-testid="review-item"
  >
    <h6
      class="text-info mb-0 lh-1"
    >
      John Chard
    </h6>
    <div
      class="d-flex justify-content-between align-items-center mb-3"
    >
      <small
        class="text-muted"
      >
        Rating: 10
      </small>
    </div>
    <p
      class="mb-2 opacity-75 review-text"
    >
      A long time ago in a childhood not too far away...
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      Princess Leia is captured and held hostage by the Imperial Army as it seeks to rule the Galactic Empire. An old Jedi Knight by the name of Ben "Obi-Wan" Kenobi may just be hers, and the rebels onl
      <span
        class="text-muted fw-bold"
      >
        ...
      </span>
    </p>
    <div
      class="d-flex mt-1 opacity-100"
    >
      <button
        class="lh-1 py-0 pb-1 rounded-0 text-decoration-none btn btn-info"
        type="button"
      >
        <small
          class="fw-bold"
          style="font-size: 13px;"
        >
          Show more
        </small>
      </button>
    </div>
  </li>
  <li
    class="px-0 bg-transparent text-white border-0 border-bottom border-secondary list-group-item"
    data-testid="review-item"
  >
    <h6
      class="text-info mb-0 lh-1"
    >
      gastyny
    </h6>
    <div
      class="d-flex justify-content-between align-items-center mb-3"
    >
      <small
        class="text-muted"
      >
        Rating: 10
      </small>
    </div>
    <p
      class="mb-2 opacity-75 review-text"
    >
      Star Wars (1977) is a true masterpiece of cinema, and is 
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
        definitely one of the best films ever made.
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      For me Star Wars (1977) is the best movie of all time,tied with The Lord of the Rings trilogy. Star Wars (1977) it is for sure the most iconic
      <span
        class="text-muted fw-bold"
      >
        ...
      </span>
    </p>
    <div
      class="d-flex mt-1 opacity-100"
    >
      <button
        class="lh-1 py-0 pb-1 rounded-0 text-decoration-none btn btn-info"
        type="button"
      >
        <small
          class="fw-bold"
          style="font-size: 13px;"
        >
          Show more
        </small>
      </button>
    </div>
  </li>
</ul>
`;
